frame/nis/src/lib.rs

```diff
@@ -0,0 +1,925 @@
+			let mut remaining = target.saturating_sub(summary.proportion_owed) * issuance.effective;
+
+			let mut queues_hit = 0;
+			let mut bids_hit = 0;
+			let mut totals = QueueTotals::<T>::get();
+			let queue_count = T::QueueCount::get();
+			totals.bounded_resize(queue_count as usize, (0, Zero::zero()));
+			for duration in (1..=queue_count).rev() {
+				if totals[duration as usize - 1].0.is_zero() {
+					continue
+				}
+				if remaining.is_zero() || queues_hit >= max_queues
+					|| !weight.check_accrue(T::WeightInfo::process_queue())
+					// No point trying to process a queue if we can't process a single bid.
+					|| !weight.can_accrue(T::WeightInfo::process_bid())
+				{
+					break
+				}
+
+				let b = Self::process_queue(
+					duration,
+					now,
+					&our_account,
+					&issuance,
+					max_bids.saturating_sub(bids_hit),
+					&mut remaining,
+					&mut totals[duration as usize - 1],
+					&mut summary,
+					weight,
+				);
+
+				bids_hit.saturating_accrue(b);
+				queues_hit.saturating_inc();
+			}
+			QueueTotals::<T>::put(&totals);
+			Summary::<T>::put(&summary);
+		}
+
+		pub(crate) fn process_queue(
+			duration: u32,
+			now: T::BlockNumber,
+			our_account: &T::AccountId,
+			issuance: &IssuanceInfo<BalanceOf<T>>,
+			max_bids: u32,
+			remaining: &mut BalanceOf<T>,
+			queue_total: &mut (u32, BalanceOf<T>),
+			summary: &mut SummaryRecordOf<T>,
+			weight: &mut WeightCounter,
+		) -> u32 {
+			let mut queue: BoundedVec<BidOf<T>, _> = Queues::<T>::get(&duration);
+			let expiry = now.saturating_add(T::Period::get().saturating_mul(duration.into()));
+			let mut count = 0;
+
+			while count < max_bids &&
+				!queue.is_empty() &&
+				!remaining.is_zero() &&
+				weight.check_accrue(T::WeightInfo::process_bid())
+			{
+				let bid = match queue.pop() {
+					Some(b) => b,
+					None => break,
+				};
+				if let Some(bid) = Self::process_bid(
+					bid,
+					expiry,
+					our_account,
+					issuance,
+					remaining,
+					&mut queue_total.1,
+					summary,
+				) {
+					queue.try_push(bid).expect("just popped, so there must be space. qed");
+				}
+				count.saturating_inc();
+			}
+			queue_total.0 = queue.len() as u32;
+			Queues::<T>::insert(&duration, &queue);
+			count
+		}
+
+		pub(crate) fn process_bid(
+			mut bid: BidOf<T>,
+			expiry: T::BlockNumber,
+			our_account: &T::AccountId,
+			issuance: &IssuanceInfo<BalanceOf<T>>,
+			remaining: &mut BalanceOf<T>,
+			queue_amount: &mut BalanceOf<T>,
+			summary: &mut SummaryRecordOf<T>,
+		) -> Option<BidOf<T>> {
+			let result = if *remaining < bid.amount {
+				let overflow = bid.amount - *remaining;
+				bid.amount = *remaining;
+				Some(Bid { amount: overflow, who: bid.who.clone() })
+			} else {
+				None
+			};
+			let amount = bid.amount.saturating_sub(T::Currency::unreserve(&bid.who, bid.amount));
+			if T::Currency::transfer(&bid.who, &our_account, amount, AllowDeath).is_err() {
+				return result
```

Comment:
It should be impossible to not be able to transfer after unreserving the amount, and there's little other way to handle it except bailing. We don't "keep the bid around", the only thing we might keep around is the remainder, in the case that the bid was too big. We have to do this because the remainder is not (yet) unreserved.

It should also be impossible for `bid.amount` to not be completely unreserved since we know that we reserved it in the first place. However it handles this circumstance gracefully by lowering the bid amount to the amount which could actually be unreserved.